File: frontend/lib/widgets/question_item.dart
import 'package:flutter/material.dart';
import '../thinker_api.dart';
import '../thinker_api_extensions.dart';

class SwipeableItem extends StatefulWidget {
  final String id;
  final Widget child;
  final Widget? background;
  final Widget? secondaryBackground;
  final Function(DismissDirection direction)? onDismissed;
  final Color? backgroundColor;
  final Color? secondaryBackgroundColor;

  const SwipeableItem({
    super.key,
    required this.id,
    required this.child,
    this.background,
    this.secondaryBackground,
    this.onDismissed,
    this.backgroundColor,
    this.secondaryBackgroundColor,
  });

  @override
  SwipeableItemState createState() => SwipeableItemState();
}

class SwipeableItemState extends State<SwipeableItem> {
  @override
  Widget build(BuildContext context) {
    return Dismissible(
      key: Key(widget.id),
      background: widget.background ?? (widget.backgroundColor != null 
        ? Container(color: widget.backgroundColor!, alignment: Alignment.centerLeft) 
        : const SizedBox.shrink()),
      secondaryBackground: widget.secondaryBackground ?? (widget.secondaryBackgroundColor != null 
        ? Container(color: widget.secondaryBackgroundColor!, alignment: Alignment.centerRight) 
        : const SizedBox.shrink()),
      onDismissed: widget.onDismissed,
      child: widget.child,
    );
  }
}

class QuestionItem extends StatelessWidget {
  final QuestionDto question;
  final String filter;
  final VoidCallback onAskLater;
  final VoidCallback onIgnore;
  final VoidCallback onUnignore;
  final VoidCallback onDeleteAnswer;
  final VoidCallback onTap;

  const QuestionItem({
    super.key,
    required this.question,
    required this.filter,
    required this.onAskLater,
    required this.onIgnore,
    required this.onUnignore,
    required this.onDeleteAnswer,
    required this.onTap,
  });

  @override
  Widget build(BuildContext context) {
    final isIgnored = question.isIgnored;
    final hasAnswer = question.isAnswered;

    Widget? background;
    Widget? secondaryBackground;
    Function(DismissDirection direction)? onDismissed;

    if (filter == 'Unanswered') {
      background = Container(
        color: Colors.blue,
        alignment: Alignment.centerLeft,
        padding: const EdgeInsets.symmetric(horizontal: 20),
        child: const Row(
          children: [
            Icon(Icons.rotate_left, color: Colors.white),
            SizedBox(width: 8),
            Text('Ask Later', style: TextStyle(color: Colors.white)),
          ],
        ),
      );
      secondaryBackground = Container(
        color: Colors.red,
        alignment: Alignment.centerRight,
        padding: const EdgeInsets.symmetric(horizontal: 20),
        child: const Row(
          mainAxisAlignment: MainAxisAlignment.end,
          children: [
            Text('Ignore', style: TextStyle(color: Colors.white)),
            SizedBox(width: 8),
            Icon(Icons.visibility_off, color: Colors.white),
          ],
        ),
      );
      onDismissed = (direction) {
        if (direction == DismissDirection.startToEnd) {
          onAskLater();
        } else if (direction == DismissDirection.endToStart) {
          onIgnore();
        }
      };
    } else if (filter == 'Answered') {
      background = Container(
        color: Colors.grey,
        alignment: Alignment.centerLeft,
        padding: const EdgeInsets.symmetric(horizontal: 20),
        child: const Row(
          children: [
            Icon(Icons.visibility_off, color: Colors.white),
            SizedBox(width: 8),
            Text('Ignore', style: TextStyle(color: Colors.white)),
          ],
        ),
      );
      secondaryBackground = Container(
        color: Colors.red,
        alignment: Alignment.centerRight,
        padding: const EdgeInsets.symmetric(horizontal: 20),
        child: const Row(
          mainAxisAlignment: MainAxisAlignment.end,
          children: [
            Text('Delete Answer', style: TextStyle(color: Colors.white)),
            SizedBox(width: 8),
            Icon(Icons.delete, color: Colors.white),
          ],
        ),
      );
      onDismissed = (direction) {
        if (direction == DismissDirection.startToEnd) {
          onIgnore();
        } else if (direction == DismissDirection.endToStart) {
          onDeleteAnswer();
        }
      };
    } else if (filter == 'Ignored') {
      background = Container(
        color: Colors.green,
        alignment: Alignment.centerLeft,
        padding: const EdgeInsets.symmetric(horizontal: 20),
        child: const Row(
          children: [
            Icon(Icons.visibility, color: Colors.white),
            SizedBox(width: 8),
            Text('Unignore', style: TextStyle(color: Colors.white)),
          ],
        ),
      );
      secondaryBackground = Container(
        color: Colors.green,
        alignment: Alignment.centerRight,
        padding: const EdgeInsets.symmetric(horizontal: 20),
        child: const Row(
          mainAxisAlignment: MainAxisAlignment.end,
          children: [
            Text('Unignore', style: TextStyle(color: Colors.white)),
            SizedBox(width: 8),
            Icon(Icons.visibility, color: Colors.white),
          ],
        ),
      );
      onDismissed = (direction) => onUnignore();
    }

    return SwipeableItem(
      id: question.id,
      background: background,
      secondaryBackground: secondaryBackground,
      onDismissed: onDismissed,
      child: Card(
        margin: const EdgeInsets.symmetric(horizontal: 16, vertical: 8),
        child: ListTile(
          title: Text(question.text),
          subtitle: hasAnswer 
            ? Text(question.currentAnswer!.text, maxLines: 1, overflow: TextOverflow.ellipsis)
            : (isIgnored ? const Text('Ignored') : null),
          trailing: Row(
            mainAxisSize: MainAxisSize.min,
            children: [
              if (filter == 'Unanswered' && !hasAnswer && !isIgnored)
                Tooltip(
                  message: 'Ask later',
                  child: IconButton(
                    icon: const Icon(Icons.rotate_left),
                    onPressed: onAskLater,
                  ),
                ),
              Tooltip(
                message: isIgnored ? 'Show question' : 'Ignore question',
                child: IconButton(
                  icon: Icon(isIgnored ? Icons.visibility : Icons.visibility_off),
                  onPressed: isIgnored ? onUnignore : onIgnore,
                ),
              ),
            ],
          ),
          onTap: onTap,
        ),
      ),
    );
  }
}

File: composeApp/src/commonMain/kotlin/alphainterplanetary/thinker/ui/components/QuestionItem.kt
package alphainterplanetary.thinker.ui.components

import alphainterplanetary.thinker.model.Question
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.RotateLeft
import androidx.compose.material.icons.filled.Edit
import androidx.compose.material.icons.filled.Visibility
import androidx.compose.material.icons.filled.VisibilityOff
import androidx.compose.material3.Card
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp

@Composable
fun QuestionItem(
  question: Question,
  filter: QuestionFilter,
  onAnswerClick: () -> Unit,
  onAskLater: () -> Unit,
  onIgnore: () -> Unit,
  onUnignore: () -> Unit,
) {
  Card(
    modifier = Modifier
      .fillMaxWidth()
      .padding(horizontal = 16.dp),
    onClick = onAnswerClick
  ) {
    Column(
      modifier = Modifier.padding(16.dp)
    ) {
      Text(
        text = question.text,
        style = MaterialTheme.typography.bodyLarge
      )

      if (question.currentAnswer != null) {
        Spacer(modifier = Modifier.height(8.dp))
        if (question.currentAnswer!!.isDraft) {
          Row(verticalAlignment = Alignment.CenterVertically) {
            Icon(
              Icons.Default.Edit,
              contentDescription = "Draft",
              modifier = Modifier.height(14.dp),
              tint = MaterialTheme.colorScheme.onSurfaceVariant,
            )
            Spacer(modifier = Modifier.width(4.dp))
            Text(
              text = "Draft:",
              style = MaterialTheme.typography.labelSmall,
              color = MaterialTheme.colorScheme.onSurfaceVariant,
            )
          }
        }
        Text(
          text = question.currentAnswer!!.text.replace(Regex("\\s+"), " ").trim(),
          maxLines = 2,
          overflow = TextOverflow.Ellipsis,
          style = MaterialTheme.typography.bodyMedium,
          color = MaterialTheme.colorScheme.onSurfaceVariant
        )
      } else if (question.isIgnored) {
        Spacer(modifier = Modifier.height(8.dp))
        Text(
          text = "Ignored",
          style = MaterialTheme.typography.bodySmall,
          color = MaterialTheme.colorScheme.error
        )
      }

      Spacer(modifier = Modifier.height(8.dp))

      Row(
        horizontalArrangement = Arrangement.End,
        modifier = Modifier.fillMaxWidth()
      ) {
        if (filter == QuestionFilter.Unanswered && !question.isAnswered && !question.isIgnored) {
          IconButton(onClick = onAskLater) {
            Icon(Icons.AutoMirrored.Filled.RotateLeft, contentDescription = "Ask later")
          }
          IconButton(onClick = onIgnore) {
            Icon(Icons.Default.VisibilityOff, contentDescription = "Ignore")
          }
        } else if (filter == QuestionFilter.Answered || filter == QuestionFilter.Ignored) {
          IconButton(onClick = if (question.isIgnored) onUnignore else onIgnore) {
            Icon(
              if (question.isIgnored) Icons.Default.Visibility else Icons.Default.VisibilityOff,
              contentDescription = if (question.isIgnored) "Unignore" else "Ignore"
            )
          }
        }
      }
    }
  }
}
